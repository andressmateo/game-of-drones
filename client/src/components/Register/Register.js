import React, { Component } from 'react';
import { Redirect } from 'react-router-dom';

import { Card, Button, LinkButton } from '../StyledComponents';
import { statusTypes } from '../../utils';

const Container = Card.extend`
  width: 500px;
  height: auto;
  input {
    font-size: 13px;
    padding: 7px 11px;
    border: solid 1px #d3dbe7;
    border-radius: 3px;
  }
  form {
    display: flex;
    flex-direction: column;
    margin: 30px 0;
    justify-content: space-between;
    height: auto;
    span {
      display: flex;
      justify-content: space-evenly;
      align-items: baseline;
      margin-top: 20px;
    }
    h5 {
      color: red;
    }
  }
`;

class Register extends Component {
  constructor(props) {
    super(props);
    this.state = { player1: '', player2: '', error: '' };
  }

  handleSubmit = event => {
    event.preventDefault();
    const { player1, player2 } = this.state;
    if (player1 && player2) {
      const { createPlayers } = this.props;
      if (createPlayers) {
        createPlayers([{ name: player1 }, { name: player2 }]);
      }
    } else {
      this.setState({ error: 'Please enter the name of each player' });
    }
  };

  render() {
    const { createPlayersStatus } = this.props;
    const { error } = this.state;
    const redirect =
      createPlayersStatus !== statusTypes.SUCCEED_STATE ? (
        ''
      ) : (
        <Redirect to="/game" />
      );
    return (
      <Container>
        {redirect}
        <h1>Enter player's names</h1>
        <form onSubmit={this.handleSubmit}>
          <span>
            <label>Player 1</label>
            <input
              type="text"
              onChange={e => this.setState({ player1: e.target.value })}
            />
          </span>
          <span>
            <label>Player 2</label>
            <input
              type="text"
              onChange={e => this.setState({ player2: e.target.value })}
            />
          </span>
          <h5>{error}</h5>
          <span>
            <Button type="submit">Start</Button>
            <LinkButton to="/players">View All Players</LinkButton>
          </span>
        </form>
      </Container>
    );
  }
}
export default Register;
