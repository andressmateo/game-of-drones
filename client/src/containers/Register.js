import { connect } from 'react-redux';

import { actions, selectors } from '../store/players';

import Register from '../components/Register';

function mapStateToProps(state) {
  const players = selectors(state);

  return {
    createPlayersStatus: players.getcreatePlayersStatus
  };
}

export default connect(mapStateToProps, {
  createPlayers: actions.create
})(Register);
