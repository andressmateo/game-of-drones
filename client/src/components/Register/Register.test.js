import React from 'react';
import { shallow } from 'enzyme';
import chai, { expect } from 'chai';
import renderer from 'react-test-renderer';
import chaiEnzyme from 'chai-enzyme';
import { MemoryRouter } from 'react-router-dom';
import chaiJestSnapshot from 'chai-jest-snapshot';
import sinon from 'sinon';

import Register from './Register';

chai.use(chaiEnzyme());
chai.use(chaiJestSnapshot);

describe('Register component', () => {
  it('renders without crashing', () => {
    shallow(<Register />);
  });

  it('renders correctly - snapshot', () => {
    const tree = renderer.create(
      <MemoryRouter>
        <Register />
      </MemoryRouter>
    );
    expect(tree).to.matchSnapshot();
  });

  it('Save user input in state', () => {
    const wrapper = shallow(<Register />);
    expect(wrapper).to.have.state('player1', '');
    wrapper
      .find('#player-one')
      .simulate('change', { target: { value: 'player one' } });
    expect(wrapper).to.have.state('player1', 'player one');
    expect(wrapper).to.have.state('player2', '');
    wrapper
      .find('#player-two')
      .simulate('change', { target: { value: 'player two' } });
    expect(wrapper).to.have.state('player2', 'player two');
  });

  it('Show an error when some name is not provided', () => {
    const wrapper = shallow(<Register />);
    const errorMessage = 'Please enter the name of each player';
    expect(wrapper).to.not.have.state('error', errorMessage);
    wrapper
      .find('#player-one')
      .simulate('change', { target: { value: 'player' } });
    wrapper
      .find('#register-form')
      .simulate('submit', { preventDefault: () => {} });
    expect(wrapper).to.have.state('error', errorMessage);
  });

  it('Calls `createPlayers` when both player`s names are provided ', () => {
    const createPlayersSpy = sinon.spy();
    const wrapper = shallow(<Register createPlayers={createPlayersSpy} />);
    wrapper
      .find('#player-one')
      .simulate('change', { target: { value: 'player1' } });
    wrapper
      .find('#player-two')
      .simulate('change', { target: { value: 'player2' } });
    wrapper
      .find('#register-form')
      .simulate('submit', { preventDefault: () => {} });
    expect(createPlayersSpy.calledOnce).to.be.true;
    expect(
      createPlayersSpy.lastCall.calledWith([
        { name: 'player1' },
        { name: 'player2' }
      ])
    ).to.be.true;
  });
});
