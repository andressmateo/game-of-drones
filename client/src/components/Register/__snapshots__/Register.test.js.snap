// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register component renders correctly - snapshot 1`] = `
<div
  className="sc-bxivhb uHiEd"
>
  
  <h1>
    Enter player's names
  </h1>
  <form
    id="register-form"
    onSubmit={[Function]}
  >
    <span>
      <label>
        Player 1
      </label>
      <input
        id="player-one"
        onChange={[Function]}
        type="text"
      />
    </span>
    <span>
      <label>
        Player 2
      </label>
      <input
        id="player-two"
        onChange={[Function]}
        type="text"
      />
    </span>
    <h5>
      
    </h5>
    <span>
      <button
        className="sc-bwzfXH fMRgfY"
        type="submit"
      >
        Start
      </button>
      <a
        className="sc-htpNat EgnXO"
        href="/players"
        onClick={[Function]}
      >
        View All Players
      </a>
    </span>
  </form>
</div>
`;
